ethereum/handler/ethereumHandler.go

```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
 }
 
 // Connect creates a new connection to the Ethereum client
-func (eh *EthereumHandler) Connect(dialEndpoint string) error {
-	if eh.EthereumClient != nil {
-		eh.EthereumClient.Close()
-	}
-	if eh.EthereumRPC != nil {
-		eh.EthereumRPC.Close()
-	}
-	maxtries := 10
+func (eh *EthereumHandler) Connect(dialEndpoints []string) error {
 	var err error
-	for {
-		if maxtries == 0 {
-			return fmt.Errorf("could not connect to web3 endpoint")
-		}
-		eh.EthereumRPC, err = ethrpc.Dial(dialEndpoint)
-		if err != nil || eh.EthereumRPC == nil {
-			log.Warnf("cannot create an ethereum rpc connection: (%s), trying again ...", err)
-			time.Sleep(time.Second * 2)
-			maxtries--
-			continue
+	for _, endpoint := range dialEndpoints {
+		maxtries := 5
+		for {
+			if maxtries == 0 {
+				log.Warnf("could not connect to %s endpoint, trying the next one ...", endpoint)
+				break
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
+			if err != nil || eh.EthereumRPC == nil {
+				log.Warnf("cannot create an ethereum rpc connection with : (%s), trying again ...", err)
+				time.Sleep(time.Second * 5)
+				maxtries--
+				continue
+			}
+			// if RPC connection established, create an ethereum client using the RPC client
+			eh.EthereumClient = ethclient.NewClient(eh.EthereumRPC)
```

Comment:
are you sure that the EthereumClient field isn't racy? it's used from other goroutines as far as I can tell.

Reply:
yes it is, adding a TODO for having 2 different connections:
- One for the subscription
- One for the ethereum sync status

Reply:
Hm, OK. Seems like a pretty big TODO, but I guess "it usually works" is better than delaying this PR.